src/rules/no_eval.rs

```diff
@@ -59,23 +62,52 @@ impl<'c> NoEvalVisitor<'c> {
   fn new(context: &'c mut Context) -> Self {
     Self { context }
   }
+
+  fn maybe_add_diagnostic(&mut self, source: &str, span: Span) {
+    if source.contains("eval") {
+      self.add_diagnostic(span);
+    }
+  }
+
+  fn add_diagnostic(&mut self, span: Span) {
+    self.context.add_diagnostic_with_hint(
+      span,
+      "no-eval",
+      "`eval` call is not allowed",
+      "Remove the use of `eval`",
+    );
+  }
 }
 
 impl<'c> Visit for NoEvalVisitor<'c> {
   noop_visit_type!();
 
+  fn visit_var_declarator(&mut self, v: &VarDeclarator, _: &dyn Node) {
+    if let Some(expr) = &v.init {
+      if let Expr::Ident(ident) = expr.as_ref() {
+        let ident_name = &ident.sym.as_ref();
+        self.maybe_add_diagnostic(ident_name, v.span);
+      }
+    }
+  }
+
   fn visit_call_expr(&mut self, call_expr: &CallExpr, _parent: &dyn Node) {
     if let ExprOrSuper::Expr(expr) = &call_expr.callee {
-      if let Expr::Ident(ident) = expr.as_ref() {
-        let name = ident.sym.as_ref();
-        if name == "eval" {
-          self.context.add_diagnostic_with_hint(
-            call_expr.span,
-            "no-eval",
-            "`eval` call is not allowed",
-            "Remove the use of `eval`",
-          );
+      match expr.as_ref() {
+        Expr::Ident(ident) => {
+          let ident_name = &ident.sym.as_ref();
+          self.maybe_add_diagnostic(ident_name, call_expr.span)
+        }
+        Expr::Member(member_expr) => {
+          let member_name = &member_expr.get_key().unwrap();
+          self.maybe_add_diagnostic(member_name, call_expr.span)
+        }
+        Expr::Paren(paren) => {
+          let paren_snippet =
+            &self.context.source_map.span_to_snippet(paren.span).unwrap();
```

Comment:
Couldn't figure out how to extract identifiers in the parens, let me know if there's a better way to do this!

Reply:
`paren` in `Expr::Paren(paren)` is `swc_ecma_ast::ParenExpr` type, which has `expr` field in it. This `expr` field is `Box<Expr>` type, and if it is equal to `eval`, it means that this `expr` looks like `Box<Expr::Ident>` where the field `sym`of `Expr::Ident` is `eval`.
Also you should probably take care of nested parens i.e. `((eval))();`

Reply:
For debugging, you may find `dbg!` so helpful. For example, if we add the following code in `Visit` implementation:

```rust
  fn visit_program(&mut self, p: &swc_ecmascript::ast::Program, _: &dyn Node) {
    dbg!(p);
    p.visit_children_with(self);
  }
```

then we can see what AST looks like.

```
[src/rules/no_eval.rs:69] p = Script(
    Script {
        span: Span {
            lo: BytePos(
                0,
            ),
            hi: BytePos(
                29,
            ),
            ctxt: #0,
        },
        body: [
            Expr(
                ExprStmt {
                    span: Span {
                        lo: BytePos(
                            0,
                        ),
                        hi: BytePos(
                            29,
                        ),
                        ctxt: #0,
                    },
                    expr: Call(
                        CallExpr {
                            span: Span {
                                lo: BytePos(
                                    0,
                                ),
                                hi: BytePos(
                                    29,
                                ),
                                ctxt: #0,
                            },
                            callee: Expr(
                                Paren(
                                    ParenExpr {
                                        span: Span {
                                            lo: BytePos(
                                                0,
                                            ),
                                            hi: BytePos(
                                                9,
                                            ),
                                            ctxt: #0,
                                        },
                                        expr: Seq(
                                            SeqExpr {
                                                span: Span {
                                                    lo: BytePos(
                                                        1,
                                                    ),
                                                    hi: BytePos(
                                                        8,
                                                    ),
                                                    ctxt: #0,
                                                },
                                                exprs: [
                                                    Lit(
                                                        Num(
                                                            Number {
                                                                span: Span {
                                                                    lo: BytePos(
                                                                        1,
                                                                    ),
                                                                    hi: BytePos(
                                                                        2,
                                                                    ),
                                                                    ctxt: #0,
                                                                },
                                                                value: 0.0,
                                                            },
                                                        ),
                                                    ),
                                                    Ident(
                                                        Ident {
                                                            span: Span {
                                                                lo: BytePos(
                                                                    4,
                                                                ),
                                                                hi: BytePos(
                                                                    8,
                                                                ),
                                                                ctxt: #1,
                                                            },
                                                            sym: Atom('eval' type=static),
                                                            type_ann: None,
                                                            optional: false,
                                                        },
                                                    ),
                                                ],
                                            },
                                        ),
                                    },
                                ),
                            ),
                            args: [
                                ExprOrSpread {
                                    spread: None,
                                    expr: Lit(
                                        Str(
                                            Str {
                                                span: Span {
                                                    lo: BytePos(
                                                        10,
                                                    ),
                                                    hi: BytePos(
                                                        28,
                                                    ),
                                                    ctxt: #0,
                                                },
                                                value: Atom('console.log(42);' type=dynamic),
                                                has_escape: false,
                                            },
                                        ),
                                    ),
                                },
                            ],
                            type_args: None,
                        },
                    ),
                },
            ),
        ],
        shebang: None,
    },
)
```

Reply:
That's extremely useful. Thanks a lot for the help!